profile-lib/utils.rkt

```diff
@@ -74,20 +74,27 @@
     (cons node (mcons (sum node-callers edge-caller edge-callee-time)
                       (sum node-callees edge-callee edge-caller-time))))
   (define nodes+io-times
-    (let loop ([todo (list root)] [r '()])
-      (if (pair? todo)
-        (let* ([cur (car todo)] [todo (cdr todo)]
-               [r (if (eq? cur root) r (cons (get-node+io cur) r))])
-          (loop (append todo ; append new things in the end, so it's a BFS
-                        (filter-map (λ (e)
-                                      (define lee (edge-callee e))
-                                      (and (not (memq lee todo))
-                                           (not (assq lee r))
-                                           lee))
-                                    (node-callees cur)))
-                r))
-        ;; note: the result does not include the root node
-        r)))
+    (let ([rhash (make-hasheq)] [todohash (make-hasheq (list (cons root #t)))])
+      (let loop ([todo (list root)] [r '()])
+        (if (pair? todo)
+          (let* ([cur (car todo)] [todo (cdr todo)]
+                 [r (if (eq? cur root) r (cons (get-node+io cur) r))])
+            (hash-remove! todohash cur)
+            (unless (eq? cur root) (hash-set! rhash cur #t))
+            (define new-things
+              (filter-map (λ (e)
```

Comment:
Can't we fuse these two loops, as in (untested):

```
(define new-things
  (for/list ([e (in-list (node-callees cur))]
             [lee (in-value (edge-callee e))]
             #:unless (hash-has-key? todohash lee)
             #:unless (hash-has-key? rhash lee))
      (hash-set! todohash lee #t)
      lee))
```

Reply:
Yeah I think so. I was trying to make minimal changes to speed it up but it probably makes more sense to rewrite to be cleaner, especially if this code ~doesn't have a maintainer. The code is written in a very Scheme, not Racket style.

I'll try to make further improvements along this line before we merge.